plugins/glitch/selectors.py

```diff
@@ -0,0 +1,107 @@
+import random
+
+#
+# Scaffolding
+#
+
+SELECTOR_MAP = {}  # TODO: add some custom error handling for missing keys
+
+def selector(func):
+    '''
+    Add a function to our dict of selectors
+
+    '''
+    SELECTOR_MAP[func.__name__] = func
+    return func
+
+#
+# Define your selectors here
+#
+
+@selector
+def units(model, units=None, applications=None):
+    '''
+    Return units that are part of the specified application(s).
+
+    If no application is specified, simply return all units.
+
+    '''
+    # Handle the case where we already have a set of units
+    if units:
+        if applications:
+            return [u for u in units if u.application in applications]
+        # noop
+        return units
+
+    # Handle the case where we do not yet have a set of units.
+    if applications is None:
+        apps = model.applications
+    else:
+        apps = [model.applications[a] for a in applications]
+
+    units = []
+    for a in apps:
+        units += a.units()
+
+    return units
+
+
+@selector
+def leader(model, units=None, value=True):
+    '''
+    Return just the units that are, or are not the leader, depending
+    on whether 'value' is truthy or falsy.
+
+    '''
+    units = units or model.units
```

Comment:
Did a bit refactor of selectors to address this.

Each selector now specifies what type of object it accepts and what type of object it returns. The objects get passed properly everywhere. And I expanded the selector map into a class that can do validation that chains of selectors actually work together.